src/goose/brokers/redis/batch.clj

```diff
@@ -0,0 +1,33 @@
+(ns goose.brokers.redis.batch
+  (:require
+    [goose.brokers.redis.commands :as redis-cmds]
+    [goose.retry :as retry]))
+
+(def ^:no-doc batch-prefix "goose/batch")
+
+(defn callback-set [batch-id] (str batch-prefix "/" batch-id))
+(defn enqueued-set [batch-id] (str batch-prefix "/" batch-id ":enqueued"))
+(defn successful-set [batch-id] (str batch-prefix "/" batch-id ":successful"))
+(defn failed-set [batch-id] (str batch-prefix "/" batch-id ":failed"))
+
+(defn update-stats [next]
+  (fn [{:keys [redis-conn]
+        :as   opts}
+       {:keys [id, batch-id]} job]
+    (if (contains? job :batch-id)
+      (try
+        (let [response (next opts job)
+              _ (redis-cmds/move-in-set redis-conn
+                                        (enqueued-set batch-id)
+                                        (successful-set batch-id)
+                                        id)]
+          response)
+        (catch Exception ex
+          (let [failed-job (goose.retry/set-failed-config job ex)]
+            (when (retry/max-retries-reached? failed-job)
```

Comment:
suggestion: you do need to move the job-id from enqueued to retry set when it fails for the first time. you can't have just the when condition, needs to handle both max-retries-reached = true and false